### ISSUE TITLE
Do not offer to run `renv::snapshot()` if there are packages that are used but not instaled.

### ISSUE THREAD

**Issue by eliocamp** [rstudio/renv#1060]
If there are packages recorded in the lockfile that are not longer used **and** packages that are on the lockfile and not installed, `status()` will first list the former and suggest running `renv::snapshot()` and then list the latter and suggest running `renv::restore()` to install them. 

If the user runs those commands in the order they appear, they will not install the desired packages, because the first call to `renv::snapshot()` will remove the needed packages from the lockfile too. 

I'd suggest only showing the message if the only effect of `renv::snapshot()` would be to remove those listed packages. Another alternative would be to have a separate function like `renv::remove_unused()` that has the predictable effect of only removing unused packages from the lockfile.

**Comment by kevinushey** [comment]
The problem here is that `renv` can't know if the user has intentionally removed these packages from the project. So the onus is on the user to decide -- do I want to use `renv::snapshot()` because I really don't want to use the package anymore, or do I want to use `renv::restore()` because `renv` was "wrong" and I do need to re-install that package?

In other words, `renv` is presenting a choice, it's not recommending that the user run those commands in sequence.

**Comment by eliocamp** [comment]
The issue is that even if the user did remove the unused packages intentionally and wants to remove them, `renv::snapshot()` has the potentially unwanted extra side-effect of also removing packages that are used in the project but not installed.

Basically, let's say you have a project that used ggplot2 and data.table. You then refactored your code and removed the data.table dependency. Also, for some reason ggplot2 was removed from the library. `renv::status()` will print:

```
The following package(s) are no longer used in this project:
               _
  data.table     [1.14.2]


Use `renv::snapshot()` to remove them from the lockfile.
```

(It will actually include also all the ggplot2 dependencies.)

The (potentially novice) user can reasonably understand that `renv::snapshot()` will only have the effect of removing data.table from the lockfile and not suspect that it will also remove ggplot2 and its dependencies. 
Below that message the user will see
```
The following package(s) are recorded in the lockfile, but not installed:
          _
  ggplot2   [tidyverse/ggplot2@HEAD]

Use `renv::restore()` to install these packages.
```

Which can clue them that there's more going on. However, the way it's framed these are presented as two independent issues which are solved via the provided commands. 

This is compounded by the fact that running `renv::status()` on a freshly-cloned project incorrectly offers the user to remove all package dependencies. This is what I saw on the wild and which prompted me to open this issue. 

1. Create a new project
2. Create a script with a `library(ggplot2)` call.
3. Run `renv::init()`
4. Delete your library to simulate someone cloning your project. (Or, clone [this minimal example project](https://github.com/eliocamp/renvtest))
5. Restart R and then use `renv::status()`. 

You will see

```
The following package(s) are no longer used in this project:
               _
  MASS           [7.3-56]
  Matrix         [1.4-1]
  R6             [2.5.1]
  RColorBrewer   [1.1-3]
  cli            [3.3.0]
  colorspace     [2.0-3]
  fansi          [1.0.3]
  farver         [2.1.1]
  glue           [1.6.2]
  gtable         [0.3.0]
  isoband        [0.2.5]
  labeling       [0.4.2]
  lattice        [0.20-45]
  lifecycle      [1.0.1]
  magrittr       [2.0.3]
  mgcv           [1.8-40]
  munsell        [0.5.0]
  nlme           [3.1-157]
  pillar         [1.8.0]
  pkgconfig      [2.0.3]
  rlang          [1.0.4]
  scales         [1.2.0]
  tibble         [3.1.8]
  utf8           [1.2.2]
  vctrs          [0.4.1]
  viridisLite    [0.4.0]
  withr          [2.5.0]

Use `renv::snapshot()` to remove them from the lockfile.

The following package(s) are recorded in the lockfile, but not installed:
          _
  ggplot2   [tidyverse/ggplot2@HEAD]

Use `renv::restore()` to install these packages.
```

That long list are ggplot2 dependencies! It seems like since ggplot2 is not installed, renv wrongly assumes that its dependencies are not longer used and offers the user to remove them. Of course, if the user now runs `renv::spanshot()`, it will be left with an unhelpful lockfile.

tldr:
There seem to be two compounding issues:
1. The message is confusing since it doesn't tell the user that `renv::snapshot()` will do more than just remove unused packages from the lockfile. 
2. renv gets confused and assumes that dependencies of not-installed packages are not used even when said packages are recorder in the lockfile, thereby printing the confusing "unused packages" message even when no packages were removed from the project.

**Comment by kevinushey** [comment]
Aha, thanks for the extra explanation -- I think I properly understand what's going on now. I agree this is not great -- I'll see if I can improve the situation here.

**Comment by kevinushey** [comment]
For reference, I think the offending code is here:

https://github.com/rstudio/renv/blob/a74721479861a7015bbb661ee8836fa65cfd18a8/R/status.R#L199-L204

The problem is that `renv_package_dependencies()` tries to compute package dependencies using the version(s) of packages currently installed, and that fails in the case where a package (e.g. `ggplot2`) is not installed.

In this case, I think we could examine the version of that package recorded in the lockfile, and use that to figure out what its dependencies are.

**Comment by alexpbell** [comment]
Hi @kevinushey fantastic package. Just wondering is there any update on this? I'm currently "renv onboarding" and this is complicating the introductory spiel that I'm doing when people clone into an renv-enabled test repository. Long list of "apparently unused" packages, but if you ignore that and stick to restore() you discover they are used after all - they're just dependencies. Thanks!

**Comment by kevinushey** [comment]
Thanks! I've improved this specific situation, where the project library is empty but the lockfile has packages recorded. In the development version of `renv`, you should now see:

```
> renv::status()
The following package(s) are recorded in the lockfile, but not installed:
               _
  MASS           [7.3-56]
  Matrix         [1.4-1]
  R6             [2.5.1]
  RColorBrewer   [1.1-3]
  cli            [3.3.0]
  colorspace     [2.0-3]
  fansi          [1.0.3]
  farver         [2.1.1]
  ggplot2        [tidyverse/ggplot2@HEAD]
  glue           [1.6.2]
  gtable         [0.3.0]
  isoband        [0.2.5]
  labeling       [0.4.2]
  lattice        [0.20-45]
  lifecycle      [1.0.1]
  magrittr       [2.0.3]
  mgcv           [1.8-40]
  munsell        [0.5.0]
  nlme           [3.1-157]
  pillar         [1.8.0]
  pkgconfig      [2.0.3]
  rlang          [1.0.4]
  scales         [1.2.0]
  tibble         [3.1.8]
  utf8           [1.2.2]
  vctrs          [0.4.1]
  viridisLite    [0.4.0]
  withr          [2.5.0]

Use `renv::restore()` to install these packages.
```

Does that seem reasonable?

**Comment by alexpbell** [comment]
Very reasonable. Can confirm that upgrading to the development version produces the expected behaviour on my test repo. Thanks!
